Make `@safe(exceptions=(TypeError, ValueError))` variant

This decorator will help to catch only expected errors, not all.

Hey @sobolevn, I'm working on this issue but I got a problem with typesafety tests :(

Here's the implementation:
```python
@overload
def safe(
    arg: Callable[..., _ValueType],
) -> Callable[..., ResultE[_ValueType]]:
    """Decorator to convert exception-throwing for any kind of Exception."""


@overload
def safe(
    arg: Tuple[Type[Exception], ...],
) -> Callable[
    [Callable[..., _ValueType]],
    Callable[..., ResultE[_ValueType]],
]:
    """Decorator to convert exception-throwing just for a set of Exceptions."""


def safe(  # noqa: WPS234
    arg: Union[
        Tuple[Type[Exception], ...],
        Callable[..., _ValueType],
    ],
) -> Union[
    Callable[
        [Callable[..., _ValueType]],
        Callable[..., ResultE[_ValueType]],
    ],
    Callable[..., ResultE[_ValueType]],
]:
    ...
```

`pytest` output:
```text
__________________________________________________________________________________________________________ safe_decorator_no_params __________________________________________________________________________________________________________
returns/typesafety/test_result/test_safe.yml:10: 
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output: 
E   Expected:
E     main:7: note: Revealed type is 'def () -> returns.result.Result[builtins.int, builtins.Exception]' (diff)
E   Actual:
E     main:7: note: Revealed type is 'def (*Any, **Any) -> returns.result.Result[builtins.int, builtins.Exception]' (diff)
```

It should work by the fact we have the `overload` signature correctly!

Do you have any idea of why it's not working?

It's not treating tuples correctly 😢 

```python
@safe(args=(ValueError, EOFError))
    def test() -> int:
        return 1
```

```text
E   Expected:
E     main:7: note: Revealed type is 'def () -> returns.result.Result[builtins.int, builtins.Exception]' (diff)
E   Actual:
E     main:3: error: No overload variant of "safe" matches argument type "Tuple[Type[ValueError], Type[EOFError]]" (diff)
E     main:3: note:     def [_ValueType] safe(arg: Callable[..., _ValueType]) -> Callable[..., Result[_ValueType, Exception]] (diff)
E     main:3: note:     def safe(arg: Tuple[Type[Exception], ...]) -> Callable[[Callable[..., _ValueType]], Callable[..., Result[_ValueType, Exception]]] (diff)
E     main:3: note: Possible overload variants:     (diff)
E     main:7: note: Revealed type is 'Any'          (diff)
```

Ok, now I think I have the same feelings you have about `mypy`!! 😠 

The problem is when decorator has parameters, like we trying to do here, see the example below:
```python
def decorator_argument(arg: int) -> Callable[[Callable[..., int]], Callable[..., int]]:
    def decorator(function: Callable[..., int]) -> Callable[..., int]:
        @wraps(function)
        def decorated(*args, **kwargs):
            pass
        return decorated
    return decorator
```

Test case:
```python
@decorator_argument(42)
def test() -> int:
    return 1

reveal_type(test)  # N: Revealed type is 'def () -> builtins.int'
```

Output:
```text
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output: 
E   Expected:
E     main:8: note: Revealed type is 'def () -> builtins.int' (diff)
E   Actual:
E     main:8: note: Revealed type is 'def (*Any, **Any) -> builtins.int' (diff)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -30,6 +30,7 @@ See [0Ver](https://0ver.org/).
 - Enables Pattern Matching support for `IOResult` container
 - Improves `hypothesis` plugin, now we detect
   when type cannot be constructed and give a clear error message
+- Adds the option to pass what exceptions `@safe` will handle
 
 
 ## 0.16.0

--- a/docs/pages/result.rst
+++ b/docs/pages/result.rst
@@ -99,6 +99,13 @@ use :func:`future_safe <returns.future.future_safe>` instead.
   >>> str(divide(0))
   '<Failure: division by zero>'
 
+If you want to `safe` handle only a set of exceptions:
+
+.. code:: python
+
+  >>> @safe((ZeroDivisionError,))  # It'll handle only `ZeroDivisionError`
+  ... def divide(number: int) -> float:
+  ...     return number / number
 
 FAQ
 ---

--- a/docs/pages/result.rst
+++ b/docs/pages/result.rst
@@ -103,9 +103,18 @@ If you want to `safe` handle only a set of exceptions:
 
 .. code:: python
 
-  >>> @safe((ZeroDivisionError,))  # It'll handle only `ZeroDivisionError`
+  >>> @safe(exceptions=(ZeroDivisionError,))  # Other exceptions will be raised
   ... def divide(number: int) -> float:
+  ...     if number > 10:
+  ...         raise ValueError('Too big')
   ...     return number / number
+
+  >>> assert divide(5) == Success(1.0)
+  >>> assert divide(0).failure()
+  >>> divide(15)
+  Traceback (most recent call last):
+    ...
+  ValueError: Too big
 
 FAQ
 ---

--- a/returns/result.py
+++ b/returns/result.py
@@ -526,6 +526,6 @@ def safe(  # type: ignore # noqa: WPS234, C901
     if callable(function):
         return factory(function, (Exception,))
     if isinstance(function, tuple):
-        exceptions = function
+        exceptions = function  # type: ignore
         function = None
     return lambda function: factory(function, exceptions)  # type: ignore

--- a/returns/result.py
+++ b/returns/result.py
@@ -506,7 +506,7 @@ def safe(  # type: ignore # noqa: WPS234, C901
       >>> assert isinstance(might_raise(0), Result.failure_type)
 
     In this case, only exceptions that are explicitly
-    listed are going to be catched.
+    listed are going to be caught.
 
     Similar to :func:`returns.io.impure_safe`
     and :func:`returns.future.future_safe` decorators.

--- a/returns/result.py
+++ b/returns/result.py
@@ -461,7 +461,7 @@ def safe(
     """Decorator to convert exception-throwing just for a set of Exceptions."""
 
 
-def safe(  # type: ignore # noqa: WPS234
+def safe(  # type: ignore # noqa: WPS234, C901
     function: Optional[Callable[_FuncParams, _ValueType]] = None,
     exceptions: Optional[Tuple[Type[Exception], ...]] = None,
 ) -> Union[

--- a/returns/result.py
+++ b/returns/result.py
@@ -446,14 +446,14 @@ ResultE = Result[_ValueType, Exception]
 
 @overload
 def safe(
-    arg: Callable[_FuncParams, _ValueType],
+    function: Callable[_FuncParams, _ValueType],
 ) -> Callable[_FuncParams, ResultE[_ValueType]]:
     """Decorator to convert exception-throwing for any kind of Exception."""
 
 
 @overload
 def safe(
-    arg: Tuple[Type[Exception], ...],
+    exceptions: Tuple[Type[Exception], ...],
 ) -> Callable[
     [Callable[_FuncParams, _ValueType]],
     Callable[_FuncParams, ResultE[_ValueType]],
@@ -461,11 +461,9 @@ def safe(
     """Decorator to convert exception-throwing just for a set of Exceptions."""
 
 
-def safe(  # noqa: WPS234
-    arg: Union[
-        Callable[_FuncParams, _ValueType],
-        Tuple[Type[Exception], ...],
-    ],
+def safe(  # type: ignore # noqa: WPS234
+    function: Optional[Callable[_FuncParams, _ValueType]] = None,
+    exceptions: Optional[Tuple[Type[Exception], ...]] = None,
 ) -> Union[
     Callable[_FuncParams, ResultE[_ValueType]],
     Callable[
@@ -494,21 +492,40 @@ def safe(  # noqa: WPS234
       >>> assert might_raise(1) == Success(1.0)
       >>> assert isinstance(might_raise(0), Result.failure_type)
 
+    You can also use it with explicit exception types as the first argument:
+
+    .. code:: python
+
+      >>> from returns.result import Result, Success, safe
+
+      >>> @safe(exceptions=(ZeroDivisionError,))
+      ... def might_raise(arg: int) -> float:
+      ...     return 1 / arg
+
+      >>> assert might_raise(1) == Success(1.0)
+      >>> assert isinstance(might_raise(0), Result.failure_type)
+
+    In this case, only exceptions that are explicitly
+    listed are going to be catched.
+
     Similar to :func:`returns.io.impure_safe`
     and :func:`returns.future.future_safe` decorators.
     """
     def factory(
-        function: Callable[_FuncParams, _ValueType],
-        exceptions: Tuple[Type[Exception], ...] = (Exception,),
+        inner_function: Callable[_FuncParams, _ValueType],
+        inner_exceptions: Tuple[Type[Exception], ...],
     ) -> Callable[_FuncParams, ResultE[_ValueType]]:
-        @wraps(function)
+        @wraps(inner_function)
         def decorator(*args: _FuncParams.args, **kwargs: _FuncParams.kwargs):
             try:
-                return Success(function(*args, **kwargs))
-            except exceptions as exc:
+                return Success(inner_function(*args, **kwargs))
+            except inner_exceptions as exc:
                 return Failure(exc)
         return decorator
 
-    if callable(arg):
-        return factory(arg)
-    return lambda func: factory(func, arg)
+    if callable(function):
+        return factory(function, (Exception,))
+    if isinstance(function, tuple):
+        exceptions = function
+        function = None
+    return lambda function: factory(function, exceptions)  # type: ignore

--- a/returns/result.py
+++ b/returns/result.py
@@ -9,9 +9,11 @@ from typing import (
     List,
     NoReturn,
     Optional,
+    Tuple,
     Type,
     TypeVar,
     Union,
+    overload,
 )
 
 from typing_extensions import ParamSpec, final
@@ -442,9 +444,35 @@ ResultE = Result[_ValueType, Exception]
 
 # Decorators:
 
+@overload
 def safe(
-    function: Callable[_FuncParams, _ValueType],
+    arg: Callable[_FuncParams, _ValueType],
 ) -> Callable[_FuncParams, ResultE[_ValueType]]:
+    """Decorator to convert exception-throwing for any kind of Exception."""
+
+
+@overload
+def safe(
+    arg: Tuple[Type[Exception], ...],
+) -> Callable[
+    [Callable[_FuncParams, _ValueType]],
+    Callable[_FuncParams, ResultE[_ValueType]],
+]:
+    """Decorator to convert exception-throwing just for a set of Exceptions."""
+
+
+def safe(  # noqa: WPS234
+    arg: Union[
+        Callable[_FuncParams, _ValueType],
+        Tuple[Type[Exception], ...],
+    ],
+) -> Union[
+    Callable[_FuncParams, ResultE[_ValueType]],
+    Callable[
+        [Callable[_FuncParams, _ValueType]],
+        Callable[_FuncParams, ResultE[_ValueType]],
+    ],
+]:
     """
     Decorator to convert exception-throwing function to ``Result`` container.
 
@@ -469,13 +497,18 @@ def safe(
     Similar to :func:`returns.io.impure_safe`
     and :func:`returns.future.future_safe` decorators.
     """
-    @wraps(function)
-    def decorator(
-        *args: _FuncParams.args,
-        **kwargs: _FuncParams.kwargs,
-    ) -> ResultE[_ValueType]:
-        try:
-            return Success(function(*args, **kwargs))
-        except Exception as exc:
-            return Failure(exc)
-    return decorator
+    def factory(
+        function: Callable[_FuncParams, _ValueType],
+        exceptions: Tuple[Type[Exception], ...] = (Exception,),
+    ) -> Callable[_FuncParams, ResultE[_ValueType]]:
+        @wraps(function)
+        def decorator(*args: _FuncParams.args, **kwargs: _FuncParams.kwargs):
+            try:
+                return Success(function(*args, **kwargs))
+            except exceptions as exc:
+                return Failure(exc)
+        return decorator
+
+    if callable(arg):
+        return factory(arg)
+    return lambda func: factory(func, arg)

--- a/tests/test_result/test_result_functions/test_safe.py
+++ b/tests/test_result/test_result_functions/test_safe.py
@@ -10,8 +10,14 @@ def _function(number: int) -> float:
     return number / number
 
 
-@safe(arg=(ZeroDivisionError,))
+@safe(exceptions=(ZeroDivisionError,))
 def _function_two(number: Union[int, str]) -> float:
+    assert isinstance(number, int)
+    return number / number
+
+
+@safe((ZeroDivisionError,))  # no name
+def _function_three(number: Union[int, str]) -> float:
     assert isinstance(number, int)
     return number / number
 
@@ -31,6 +37,9 @@ def test_safe_failure_with_expected_error():
     """Ensures that safe decorator works correctly for Failure case."""
     failed = _function_two(0)
     assert isinstance(failed.failure(), ZeroDivisionError)
+
+    failed2 = _function_three(0)
+    assert isinstance(failed2.failure(), ZeroDivisionError)
 
 
 def test_safe_failure_with_non_expected_error():

--- a/tests/test_result/test_result_functions/test_safe.py
+++ b/tests/test_result/test_result_functions/test_safe.py
@@ -1,9 +1,18 @@
+from typing import Union
+
+import pytest
 
 from returns.result import Success, safe
 
 
 @safe
 def _function(number: int) -> float:
+    return number / number
+
+
+@safe(arg=(ZeroDivisionError,))
+def _function_two(number: Union[int, str]) -> float:
+    assert isinstance(number, int)
     return number / number
 
 
@@ -16,3 +25,15 @@ def test_safe_failure():
     """Ensures that safe decorator works correctly for Failure case."""
     failed = _function(0)
     assert isinstance(failed.failure(), ZeroDivisionError)
+
+
+def test_safe_failure_with_expected_error():
+    """Ensures that safe decorator works correctly for Failure case."""
+    failed = _function_two(0)
+    assert isinstance(failed.failure(), ZeroDivisionError)
+
+
+def test_safe_failure_with_non_expected_error():
+    """Ensures that safe decorator works correctly for Failure case."""
+    with pytest.raises(AssertionError):
+        _function_two('0')

--- a/typesafety/test_result/test_safe.yml
+++ b/typesafety/test_result/test_safe.yml
@@ -10,6 +10,18 @@
     reveal_type(test)  # N: Revealed type is "def () -> returns.result.Result[builtins.int, builtins.Exception]"
 
 
+- case: safe_decorator_passing_exceptions_no_params
+  disable_cache: false
+  main: |
+    from returns.result import safe
+
+    @safe((ValueError,))
+    def test() -> int:
+        return 1
+
+    reveal_type(test)  # N: Revealed type is "def () -> returns.result.Result[builtins.int, builtins.Exception]"
+
+
 - case: safe_composition_no_params
   disable_cache: false
   main: |
@@ -21,6 +33,17 @@
     reveal_type(safe(test))  # N: Revealed type is "def () -> returns.result.Result[builtins.int, builtins.Exception]"
 
 
+- case: safe_composition_passing_exceptions_no_params
+  disable_cache: false
+  main: |
+    from returns.result import safe
+
+    def test() -> int:
+        return 1
+
+    reveal_type(safe((EOFError,))(test))  # N: Revealed type is "def () -> returns.result.Result[builtins.int, builtins.Exception]"
+
+
 - case: safe_decorator_with_args
   disable_cache: false
   main: |
@@ -28,6 +51,19 @@
     from returns.result import safe
 
     @safe
+    def test(first: int, second: Optional[str] = None, *, kw: bool = True) -> int:
+        return 1
+
+    reveal_type(test)  # N: Revealed type is "def (first: builtins.int, second: Union[builtins.str, None] =, *, kw: builtins.bool =) -> returns.result.Result[builtins.int, builtins.Exception]"
+
+
+- case: safe_decorator_passing_exceptions_with_args
+  disable_cache: false
+  main: |
+    from typing import Optional
+    from returns.result import safe
+
+    @safe((ValueError, EOFError))
     def test(first: int, second: Optional[str] = None, *, kw: bool = True) -> int:
         return 1
 
@@ -46,6 +82,18 @@
     reveal_type(safe(test))  # N: Revealed type is "def (first: builtins.int, second: Union[builtins.str, None] =, *, kw: builtins.bool =) -> returns.result.Result[builtins.int, builtins.Exception]"
 
 
+- case: safe_composition_passing_exceptions_with_args
+  disable_cache: false
+  main: |
+    from typing import Optional
+    from returns.result import safe
+
+    def test(first: int, second: Optional[str] = None, *, kw: bool = True) -> int:
+        return 1
+
+    reveal_type(safe((ValueError,))(test))  # N: Revealed type is "def (first: builtins.int, second: Union[builtins.str, None] =, *, kw: builtins.bool =) -> returns.result.Result[builtins.int, builtins.Exception]"
+
+
 - case: safe_regression333
   disable_cache: false
   main: |
@@ -53,6 +101,19 @@
     from typing import Any
 
     @safe
+    def send(text: str) -> Any:
+        return "test"
+
+    reveal_type(send)  # N: Revealed type is "def (text: builtins.str) -> returns.result.Result[Any, builtins.Exception]"
+
+
+- case: safe_passing_exceptions_regression333
+  disable_cache: false
+  main: |
+    from returns.result import safe
+    from typing import Any
+
+    @safe((Exception,))
     def send(text: str) -> Any:
         return "test"
 
@@ -72,12 +133,37 @@
     reveal_type(safe(tap(Response.raise_for_status)))  # N: Revealed type is "def (main.Response*) -> returns.result.Result[main.Response, builtins.Exception]"
 
 
+- case: safe_passing_exceptions_regression641
+  disable_cache: false
+  main: |
+    from returns.result import safe
+    from returns.functions import tap
+
+    class Response(object):
+        def raise_for_status(self) -> None:
+            ...
+
+    reveal_type(safe((EOFError,))(tap(Response.raise_for_status)))  # N: Revealed type is "def (main.Response*) -> returns.result.Result[main.Response, builtins.Exception]"
+
+
 - case: safe_decorator_with_args_kwargs
   disable_cache: false
   main: |
     from returns.result import safe
 
     @safe
+    def test(*args, **kwargs) -> int:
+        return 1
+
+    reveal_type(test)  # N: Revealed type is "def (*args: Any, **kwargs: Any) -> returns.result.Result[builtins.int, builtins.Exception]"
+
+
+- case: safe_decorator_passing_exceptions_with_args_kwargs
+  disable_cache: false
+  main: |
+    from returns.result import safe
+
+    @safe((EOFError,))
     def test(*args, **kwargs) -> int:
         return 1
 
@@ -96,6 +182,18 @@
     reveal_type(test)  # N: Revealed type is "def (*args: builtins.int, **kwargs: builtins.str) -> returns.result.Result[builtins.int, builtins.Exception]"
 
 
+- case: safe_decorator_passing_exceptions_with_args_kwargs
+  disable_cache: false
+  main: |
+    from returns.result import safe
+
+    @safe((Exception,))
+    def test(*args: int, **kwargs: str) -> int:
+        return 1
+
+    reveal_type(test)  # N: Revealed type is "def (*args: builtins.int, **kwargs: builtins.str) -> returns.result.Result[builtins.int, builtins.Exception]"
+
+
 - case: safe_decorator_composition
   disable_cache: false
   main: |
@@ -104,6 +202,20 @@
 
     @impure
     @safe
+    def test(*args: int, **kwargs: str) -> int:
+        return 1
+
+    reveal_type(test)  # N: Revealed type is "def (*args: builtins.int, **kwargs: builtins.str) -> returns.io.IO[returns.result.Result*[builtins.int*, builtins.Exception]]"
+
+
+- case: safe_decorator_passing_exceptions_composition
+  disable_cache: false
+  main: |
+    from returns.io import impure
+    from returns.result import safe
+
+    @impure
+    @safe((ValueError,))
     def test(*args: int, **kwargs: str) -> int:
         return 1
 

--- a/typesafety/test_result/test_safe.yml
+++ b/typesafety/test_result/test_safe.yml
@@ -21,6 +21,12 @@
 
     reveal_type(test)  # N: Revealed type is "def () -> returns.result.Result[builtins.int, builtins.Exception]"
 
+    @safe(exceptions=(ValueError,))
+    def test2() -> int:
+        return 1
+
+    reveal_type(test2)  # N: Revealed type is "def () -> returns.result.Result[builtins.int, builtins.Exception]"
+
 
 - case: safe_composition_no_params
   disable_cache: false

